src/tools/run-make-support/src/assertion_helpers/mod.rs

```diff
@@ -0,0 +1,160 @@
+//! Collection of assertions and assertion-related helpers.
+
+#[cfg(test)]
+mod tests;
+
+use std::panic;
+use std::path::Path;
+
+use crate::{fs, regex};
+
+/// Assert that `actual` is equal to `expected`.
+#[track_caller]
+pub fn assert_equals<A: AsRef<str>, E: AsRef<str>>(actual: A, expected: E) {
+    let actual = actual.as_ref();
+    let expected = expected.as_ref();
+
+    if actual != expected {
+        eprintln!("=== ACTUAL TEXT ===");
+        eprintln!("{}", actual);
+        eprintln!("=== EXPECTED ===");
+        eprintln!("{}", expected);
+        panic!("expected text does not match actual text");
+    }
+}
+
+struct SearchDetails<'assertion_name, 'haystack, 'query> {
+    assertion_name: &'assertion_name str,
+    haystack: &'haystack str,
+    query_kind: &'static str,
+    query: &'query str,
+}
+
+impl<'assertion_name, 'haystack, 'query> SearchDetails<'assertion_name, 'haystack, 'query> {
+    fn dump(&self) {
+        eprintln!("{}:", self.assertion_name);
+        eprintln!("=== HAYSTACK ===");
+        eprintln!("{}", self.haystack);
+        eprintln!("=== {} ===", self.query_kind);
```

Comment:
This is never not "NEEDLE". Why abstract it?

Reply:
See the regex variants (e.g. `assert_contains_regex` and friends), this can actually be `query_kind: "REGEX",`